metadata-ingestion/src/datahub/ingestion/source/hex/mapper.py

```diff
@@ -0,0 +1,330 @@
+        self,
+        workspace_name: str,
+        platform_instance: Optional[str] = None,
+        base_url: str = HEX_API_BASE_URL_DEFAULT,
+        patch_metadata: bool = True,
+        collections_as_tags: bool = True,
+        status_as_tag: bool = True,
+        categories_as_tags: bool = True,
+        set_ownership_from_email: bool = True,
+    ):
+        self._workspace_name = workspace_name
+        self._platform_instance = platform_instance
+        self._workspace_urn = self._get_workspace_urn(
+            workspace_name=workspace_name, platform_instance=platform_instance
+        )
+        self._base_url = base_url.strip("/").replace("/api/v1", "")
+        self._patch_metadata = patch_metadata
+        self._collections_as_tags = collections_as_tags
+        self._status_as_tag = status_as_tag
+        self._categories_as_tags = categories_as_tags
+        self._set_ownership_from_email = set_ownership_from_email
+
+    def map_workspace(self) -> Iterable[MetadataWorkUnit]:
+        container_properties = ContainerPropertiesClass(
+            name=self._workspace_name,
+        )
+        aspects: List[_Aspect] = [container_properties]
+        yield from self._yield_mcps(
+            entity_urn=self._workspace_urn,
+            aspects=aspects,
+        )
+
+    def map_project(self, project: Project) -> Iterable[MetadataWorkUnit]:
+        dashboard_urn = self._get_dashboard_urn(name=project.id)
+
+        dashboard_info = DashboardInfoClass(
+            title=project.title,
+            description=project.description or "",
+            lastModified=self._change_audit_stamps(
+                created_at=project.created_at, last_edited_at=project.last_edited_at
+            ),
+            externalUrl=f"{self._base_url}/{self._workspace_name}/hex/{project.id}",
+            customProperties={k: str(v) for k, v in project._raw_data.items()}
+            if project._raw_data
+            else None,
+        )
+
+        subtypes = SubTypesClass(
+            typeNames=[BIAssetSubTypes.HEX_PROJECT],
+        )
+
+        platform_instance = self._platform_instance_aspect()
+
+        container = ContainerClass(
+            container=self._workspace_urn.urn(),
+        )
+
+        aspects = [dashboard_info, subtypes, platform_instance, container]
+
+        tags = self._global_tags(
+            status=project.status,
+            categories=project.categories,
+            collections=project.collections,
+        )
+        if tags:
+            aspects.append(tags)
+
+        ownership = self._ownership(creator=project.creator, owner=project.owner)
+        if ownership:
+            aspects.append(ownership)
+
+        dashboard_usage_statistics = self._dashboard_usage_statistics(
+            analytics=project.analytics
+        )
+        if dashboard_usage_statistics:
+            aspects.append(dashboard_usage_statistics)
+
+        yield from self._yield_mcps(entity_urn=dashboard_urn, aspects=aspects)
```

Comment:
there's a `MCPW.construct_many` method

also, it'd be nicer if this method returned List[MCPW] - will make testing a lot easier as well. you then can use the `auto_workunit` helper in hex.py to convert them into workunits

Reply:
> there's a `MCPW.construct_many` method

yes, I'm using that one in `_yield_mcps`

> also, it'd be nicer if this method returned List[MCPW] - will make testing a lot easier as well. you then can use the `auto_workunit` helper in hex.py to convert them into workunits

I end up with current approach because I want the mapper to be the one responsible for producing the patch events, if necessary.

Since I'm using the existing function `convert_dashboard_info_to_patch`, I need to operate at `MetadataWorkUnit` level

```
def convert_dashboard_info_to_patch(
    urn: str, aspect: DashboardInfoClass, system_metadata: Optional[SystemMetadataClass]
) -> Optional[MetadataWorkUnit]:
```

Reply:
how about once this is in https://github.com/datahub-project/datahub/pull/12947

Reply:
What prevents this mapper to operate at MCPW level is `convert_dasboard_info_to_patch`, which returns WU. If that method returned MCPW, I could move mapper to MCPW too.

Of course, with that PR, I could make mapper to return `Union[MCPW|WU]`, however I don't see that as a big improvement, actually the opposite; more confusing